### DISCUSSION TITLE
Screen share not working properly within SEB with Sonoma 14.5

### DISCUSSION THREAD

**Discussion by hlo250** [SafeExamBrowser/seb-mac#413] [General]
Ever since 14.5 was released, we've been having issues with sharing the screen properly within SEB. 

When i click on share this screen, the screen that is shared is not the Safe Exam Browser screen, instead it shares my desktop. The only way to fix it, is to click on the presenter overlay "small" option and then turning it off, and then it finally shows the SEB. 

Tested on previous versions of Sonoma and screensharing doesn't have any issues. Anyone have any ideas?

![image](https://github.com/SafeExamBrowser/seb-mac/assets/171949201/2f327624-8933-4126-beb4-da0f0d970d28)

![image](https://github.com/SafeExamBrowser/seb-mac/assets/171949201/6519a58e-b472-436a-80aa-8cc06448f204)

**Comment by bhardwaj-manish** [comment]
We are facing a similar issue with SEB on macOS 14.5. When the screen is recorded using the Screen Capture API with the getDisplayMedia() method, SEB's screen is not recorded, only the desktop screen is recorded. However, mouse movements are visible. When we tried the screen recording on macOS 12.7.5, it worked fine.

After investigation, We found that the newer macOS version has added some new options in Safari to share either the whole screen or just a window, which were not present in older versions of Safari before 14.5 mac os. Since SEB also uses WebKit, this new change might be causing the issue.

Video track object returned by MediaStreamTrack: getSettings() method:

macOS v12.7.5:
```
{
    deviceId: "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    frameRate: 30,
    height: 0,
    width: 0,
}
```

macOS v14.5:
```
{
    deviceId: "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    displaySurface: "monitor",
    frameRate: 30,
    height: 0,
    width: 0,
}
```

As we can see above, displaySurface was not present in the video's settings in the older macOS version.

### Current Solution (Not Suitable for our requirements)
By tweaking SEB configurations, We found that enabling the `allowWindowCapture` setting fixes the screen recording problem. This generates a possibility that this issue might be happening due to a mismatch between SEB and the new WebKit settings. It seems SEB's security settings are now stopping it from recording its own screen.

While enabling allowWindowCapture solves the screen recording issue, it also lets users take screenshots, which is not desirable for our requirements. Please let us know how we can achieve the same result without allowing our screen to be captured by users.

Device configurations in which the issue was faced:
macOS version: 14.5
SEB version: 3.3.3
Safari version: 17.5
Webkit version: 605.1.15

**Comment by danschlet** [comment]
Apple must have fixed in macOS 14.5 that screen sharing wasn't respecting the NSWindowSharingNone property option on NSWindow. That's good for security, but bad if you attempt to use getDisplayMedia() with SEB.

But you're almost there, you can configure SEB to allow window capture but to use the legacy method to stop screen shots. Just check the setting below Allow Window Capture. [It is mentioned in the manual ](https://safeexambrowser.org/macosx/mac_usermanual_en.html#SecurityPane).

**Reply by hlo250** [reply]
@danschlet I enabled both allow window capture and block screen shots (legacy) but screenshots are still available via CMD+ Shift+ 3. All other commands seem be blocked except the one mentioned. Is there a way to prevent that command from working?

**Reply by danschlet** [reply]
What do you mean that they are still available with cmd-3? Can you see these screen shots somewhere after exiting SEB? In your desktop folder?

**Reply by hlo250** [reply]
@danschlet Yes, after exiting the SEB, the screenshots are in my desktop folder.

**Reply by danschlet** [reply]
It's possible that the legacy blocking of screen shots doesn't work in SEB 3.3.3 (or earlier), although in my case it still worked (maybe deleting SEB client settings would fix it in 3.3.3 temporarily, which on macOS Sonoma has to be done with deleting the org.safeexambrowser.Safe-Exam-Browser.plist file from the ~/Library/Preferences folder (the defaults delete terminal command doesn't work with macOS 14 anymore).

But best would be if you could try it with SEB 3.4b1. I fixed something related to legacy screen shot blocking. Also it would be very helpful if you also could try if web screen sharing still works when you (after allowing window capture) again disable "Allow screen capture/recording". I assume it won't work, which creates another security issue.

I hope Apple can provide a solution for upcoming macOS versions. Are you sure it still worked in macOS 14.0, 14.1 etc. before 14.5?

**Reply by hlo250** [reply]
@danschlet Appreciate the info, but even after using 3.4b1 and applying the proposed settings, screenshotting is still possible.

Yes, anything before 14.5 works fine.

**Reply by danschlet** [reply]
Hmm, are you using a Mac user with administrator rights?

Can you send me the SEB log file from the session you tried it (see Console app or just search for org.safeexambrowser.SafeExamBrowser in Finder)?

The legacy screen shot blocking feature definitely works for me with SEB 3.4b1 on Sonoma 14.5.

**Reply by danschlet** [reply]
It doesn't look that in the session which is covered inside this log file you had these settings active
Allow screen capture/recording
Allow window capture (screen shots)
Block screen shots (Legacy)

**Reply by hlo250** [reply]
@danschlet Apologies, here is the log: 
[org.safeexambrowser.SafeExamBrowser 2024-08-07--22-28-44-486.log](https://github.com/user-attachments/files/16534878/org.safeexambrowser.SafeExamBrowser.2024-08-07--22-28-44-486.log)

**Comment by danschlet** [comment]
This also applies to SEB's own screen recording feature, from SEB 3.4b1 release notes: Currently screen proctoring requires the used settings ([Security pane](https://safeexambrowser.org/macosx/mac_usermanual_en.html#SecurityPane)) to explicitly allow:
1. Allow screen capture/recording
2. Allow window capture (screen shots)
3. Block screen shots (Legacy)

<img width="779" alt="Screenshot 2024-08-06 at 13 30 21" src="https://github.com/user-attachments/assets/f010fbd1-ed6e-4303-9c56-2d7511b561c3">
